modules/assim.batch/R/autoburnin.R

```diff
@@ -1,47 +1,115 @@
+#' `gelman.plot` method uses the `gelman.plot` function from the `coda` package to calculate the Gelman diagnostic cumulatively from the beginning of the chain.
+#' The iteration after the last time the `gelman.diag` exceeds the threshold for any of the parameters is the burnin.
+#' 
 #' @examples
 #'      library(coda)
 #'      data(line)
-#'      burnin <- getBurnin(line, threshold = 1.05)
-#' @author Michael Dietze, Alexey Shiklomanov
+#'      burnin <- getBurnin(line, threshold = 1.05, 
+#'                          method = 'gelman.plot')
+#' @author Alexey Shiklomanov, Michael Dietze
 #' @export
-getBurnin <- function(jags_out, threshold = 1.1, use.confidence = TRUE, 
-                      autoburnin = FALSE, plotfile = "/dev/null", ...) {
-  if (length(find("logger.info")) == 0) {
-    msg <- message
+getBurnin <- function(jags_out, 
+                      method = "rmw", 
+                      threshold = 1.1,
+                      use.confidence = TRUE,
+                      width = ceiling(coda::niter(jags_out)/2), 
+                      njump = 50,
+                      autoburnin = FALSE,
+                      plotfile = "/dev/null",
+                      ...) {
+  if (!coda::is.mcmc.list(jags_out)) jags_out <- makeMCMCList(jags_out)
+  stopifnot(coda::niter(jags_out) > 50)
+  if (method == "rmw"){
+    burnin <- getBurnin.rmw(jags_out, 
+                            width = width,
+                            njump = njump,
+                            threshold = threshold,
+                            use.confidence = use.confidence,
+                            ...)
+  } else if (method == "gelman.plot") {
+    burnin <- getBurnin.gelman.plot(jags_out, 
+                                    threshold = threshold,
+                                    use.confidence = use.confidence,
+                                    autoburnin = autoburnin,
+                                    plotfile = plotfile,
+                                    ...)
   } else {
-    msg <- logger.info
+    stop(sprintf("Method '%s' not valid. Only 'rmw' and 'gelman.plot' currently supported.", 
+                method))
   }
-  if (!is.mcmc.list(jags_out)) {
-    jags_out <- makeMCMCList(jags_out)
+  if (burnin == 1) message("*** Chains have not converged yet ***")
+  return(burnin)
+}
+
+getBurnin.rmw <- function(x, width = ceiling(coda::niter(x)/2), njump = 50,
+                          threshold = 1.1, use.confidence = TRUE, ...) {
+  stopifnot(width %% 1 == 0)
+  stopifnot(njump %% 1 == 0)
+  startx <- start(x)
+  endx <- end(x)
+  converged <- TRUE
+  gdcol <- ifelse(use.confidence, 2, 1)
+  a <- floor(seq(endx - width + 1, startx, length.out = njump))
+  b <- ceiling(seq(endx, startx + width - 1, length.out = njump))
+  i <- 1
+  while(converged & i <= length(a)) {
+    xsub <- window(x, start=a[i], end=b[i])
+    gd_raw <- coda::gelman.diag(xsub, autoburnin=FALSE)
+    gd <- c(gd_raw$psrf[,gdcol], gd_raw$mpsrf)
+    if (any(gd > threshold)) {
```

Comment:
I think there is a scope issue here, `burnin` is not assigned unless this condition is met